### PR TITLE
[provider-local] Use `gardener-operator` for local development setup

### PR DESCRIPTION
**Pull request by oliver-goetz**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
Created by @robertvolkmann and me during [Gardener Hackathon 05/2024](https://github.com/gardener-community/hackathon/tree/main/2024-05_Schelklingen).

Until now, it was not possible to create seeds and shoots in the `gardener-operator` for local development setup (started via `make kind-operator-up operator-up`)
This PR adds this capability. There is a new make target `operator-seed-up` to bring up a new garden and deploys a gardenlet into the kind cluster. Thus, the complete scenario can be started with `make kind-operator-up operator-up operator-seed-up`.

There is a `kubeconfig` for the virtual-garden at `./example/operator/virtual-garden/kubeconfig`. 
The `kubeconfig` of the kind cluster is still at available at `./example/gardener-local/kind/operator/kubeconfig`.

Starting this scenario takes a longer to start then `make kind-up gardener-up` and consumes more memory. These are the values we measured. 
```
kind-operator-up operator-up operator-seed-up
startup: 10 minutes
memory: 6.4 GB

kind-ha-multi-zone-up gardener-ha-multi-zone-up
startup:  6.20 minutes
memory: 4.2 GB

kind-up gardener-up
startup: 6 minutes
memory: 3 GB
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature developer
`gardener-operator` local development setup supports creating seeds, shoots and managed-seeds now. Please see [the docs](https://github.com/gardener/gardener/blob/master/docs/deployment/getting_started_locally.md#alternative-way-to-set-up-garden-and-seed-leveraging-gardener-operator) for details.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
> Thus, the complete scenario can be started with `make kind-operator-up operator-up operator-garden-seed-up`.

What about renaming `operator-garden-seed-up` to `operator-seed-up` then?

File: Makefile [hunk 1]
```diff
@@ -296,28 +298,26 @@ kind-extensions-clean:
 
 kind-operator-up: $(KIND) $(KUBECTL) $(HELM) $(YQ) $(KUSTOMIZE)
 	./hack/kind-up.sh \
-		--cluster-name gardener-operator-local \
-		--path-kubeconfig $(REPO_ROOT)/example/gardener-local/kind/operator/kubeconfig \
+		--cluster-name $(CLUSTER_NAME) \
+		--path-kubeconfig $(KIND_KUBECONFIG) \
 		--path-cluster-values $(REPO_ROOT)/example/gardener-local/kind/operator/values.yaml \
 		--with-lpp-resize-support $(DEV_SETUP_WITH_LPP_RESIZE_SUPPORT)
-	mkdir -p $(REPO_ROOT)/dev/local-backupbuckets/gardener-operator
```
Comment (rfranzke) [inline Makefile:303]: Why this line was removed?

File: Makefile [hunk 2]
```diff
@@ -381,6 +381,17 @@ operator-down: $(SKAFFOLD) $(HELM) $(KUBECTL)
 	$(KUBECTL) delete garden --all --ignore-not-found --wait --timeout 5m
 	$(SKAFFOLD) delete
 
+operator-garden-seed-%: export VIRTUAL_GARDEN_KUBECONFIG = $(REPO_ROOT)/example/gardener-local/virtual-garden/kubeconfig
+
+operator-garden-seed-up: $(SKAFFOLD) $(HELM) $(KUBECTL)
+	$(SKAFFOLD) run -m garden -f=skaffold-operator-garden.yaml
+	$(SKAFFOLD) run -m garden-namespace -f=skaffold-operator-garden.yaml --kubeconfig=$(VIRTUAL_GARDEN_KUBECONFIG) --status-check=false # deployments doesn't exist in virtual-garden, see https://skaffold.dev/docs/status-check/
```
Comment (rfranzke) [inline Makefile]:
```suggestion
	$(SKAFFOLD) run -m garden-namespace -f=skaffold-operator-garden.yaml --kubeconfig=$(VIRTUAL_GARDEN_KUBECONFIG) --status-check=false # deployments don't exist in virtual-garden, see https://skaffold.dev/docs/status-check/
```

File: Makefile [hunk 3]
```diff
@@ -381,6 +381,17 @@ operator-down: $(SKAFFOLD) $(HELM) $(KUBECTL)
...
+	$(SKAFFOLD) run -m provider-local -p operator -f=skaffold.yaml --kubeconfig=$(VIRTUAL_GARDEN_KUBECONFIG) --status-check=false # deployments doesn't exist in virtual-garden, see https://skaffold.dev/docs/status-check/
```
Comment (rfranzke) [inline Makefile]:
```suggestion
	$(SKAFFOLD) run -m provider-local -p operator -f=skaffold.yaml --kubeconfig=$(VIRTUAL_GARDEN_KUBECONFIG) --status-check=false # deployments don't exist in virtual-garden, see https://skaffold.dev/docs/status-check/
```

File: charts/gardener/operator/templates/deployment.yaml [hunk 4]
```diff
@@ -140,10 +140,6 @@ spec:
         {{- end }}
         - name: gardener-operator-config
           mountPath: /etc/gardener-operator/config
-{{- if .Values.hostAliases }}
-      hostAliases:
-{{ toYaml .Values.hostAliases | indent 6 }}
-{{- end }}
```
Comment (rfranzke) [inline charts/gardener/operator/templates/deployment.yaml]: Why this was removed? The Helm chart is of "general purpose" (not only used for local dev), so probably we should keep allowing this configuration?

File: charts/gardener/provider-local/templates/deployment.yaml [hunk 5]
```diff
@@ -32,6 +32,7 @@ spec:
         networking.gardener.cloud/to-runtime-apiserver: allowed
         networking.gardener.cloud/to-dns: allowed
         networking.gardener.cloud/to-kind-network: allowed
+        networking.gardener.cloud/to-virtual-garden-istio: allowed
```
Comment (rfranzke) [inline charts/gardener/provider-local/templates/deployment.yaml]: Use `networking.resources.gardener.cloud/to-all-istio-ingresses-istio-ingressgateway-tcp-9443: allowed` instead and add `{"matchLabels":{"gardener.cloud/role":"extension"}}` to https://github.com/gardener/gardener/blob/dcf0191b4f2a883f9115fd8316bc6884316abdf6/pkg/component/networking/istio/charts/istio/istio-ingress/templates/service.yaml#L8

File: charts/gardener/provider-local/templates/networkpolicy.yaml [hunk 6]
```diff
@@ -19,3 +19,27 @@ spec:
...
+  - to:
+    - namespaceSelector:
+        matchLabels:
+          kubernetes.io/metadata.name: virtual-garden-istio-ingress
+      podSelector:
+        matchLabels:
+          app: istio-ingressgateway
+          istio: ingressgateway
+  policyTypes:
+  - Egress
```
Comment (rfranzke) [inline charts/gardener/provider-local/templates/networkpolicy.yaml]: delete this

File: hack/operator-garden-seed-down.sh [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+kubectl --kubeconfig "$PATH_GARDEN_KUBECONFIG" annotate project local garden confirmation.gardener.cloud/deletion=true
+skaffold -f=skaffold.yaml --kubeconfig "$PATH_KIND_KUBECONFIG" delete -m gardenlet -p operator
+
+# workaround for https://github.com/gardener/gardener/issues/5164
+kubectl --kubeconfig "$PATH_KIND_KUBECONFIG" delete ns \
+  seed-local \
+  --ignore-not-found
+
+# cleanup namespaces that don't get deleted automatically
+kubectl delete ns gardener-system-seed-lease --ignore-not-found
```
Comment (rfranzke) [inline hack/operator-garden-seed-down.sh]: Does this work at all (which kubeconfig is used)?
Anyways, this can be removed, no? You cleanup the `Garden` next, so the virtual cluster will be deleted, and with it, this namespace.

File: pkg/provider-local/local/types.go [hunk 8]
```diff
@@ -31,6 +31,11 @@ const (
 	// 'networking.gardener.cloud/to-istio-ingressgateway=allowed' to istio-ingressgateway pods running in
 	// 'istio-ingress' namespace.
 	LabelNetworkPolicyToIstioIngressGateway = "networking.gardener.cloud/to-istio-ingressgateway"
+
+	// LabelNetworkPolicyToVirtualGarden allows Egress from pods labeled with
+	// 'networking.resources.gardener.cloud/to-all-istio-ingresses-istio-ingressgateway-tcp-9443' to istio-ingressgateway
+	// pods running in the namespaces 'istio-ingress' and 'virtual-garden-istio-ingress'.
+	LabelNetworkPolicyToVirtualGarden = "networking.resources.gardener.cloud/to-all-istio-ingresses-istio-ingressgateway-tcp-9443"
```
Comment (rfranzke) [inline pkg/provider-local/local/types.go]: Remove this constant. We don't have such constants for any other services anywhere else.

File: pkg/provider-local/webhook/dnsconfig/mutator.go [hunk 9]
```diff
@@ -63,6 +63,7 @@ func (m *mutator) Mutate(ctx context.Context, newObj, oldObj client.Object) erro
 	}
 
 	metav1.SetMetaDataLabel(podMeta, local.LabelNetworkPolicyToIstioIngressGateway, v1beta1constants.LabelNetworkPolicyAllowed)
+	metav1.SetMetaDataLabel(podMeta, local.LabelNetworkPolicyToVirtualGarden, v1beta1constants.LabelNetworkPolicyAllowed)
```
Comment (rfranzke) [inline pkg/provider-local/webhook/dnsconfig/mutator.go]: This should only be added to machine pods (and BBX?), right? DWD doesn't need this, does it?

File: pkg/controller/service/reconciler.go [hunk 10]
```diff
@@ -66,29 +68,57 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+	case key == keyVirtualGardenIstioIngressGateway:
+		nodePort = nodePortVirtualGardenKubeAPIServer
+		if r.IsMultiZone {
+			// Docker desktop for mac v4.23 breaks traffic going through a port mapping to a different docker container.
+			// Setting external traffic policy to local mitigates the issue for multi-node setups, e.g. for gardener-operator.
+			service.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyLocal
+		}
+		if nodes, err := r.getNodesInternalIPs(ctx, client.MatchingLabels{"node-role.kubernetes.io/control-plane": ""}); err != nil {
+			return reconcile.Result{}, err
+		} else {
```
Comment (rfranzke) [inline pkg/controller/service/reconciler.go]: `else` not needed since you return anyways

File: skaffold-operator-garden.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        - host:
+            command:
+              - bash
+              - -ec
+              - kubectl -n garden get secret gardener -o jsonpath={.data.kubeconfig} | base64 -d > $VIRTUAL_GARDEN_KUBECONFIG
+---
+apiVersion: skaffold/v4beta7
+kind: Config
+metadata:
+  name: garden-namespace
```
Comment (rfranzke) [inline skaffold-operator-garden.yaml]: I would call this `gardener-config`

**Review by rfranzke — CHANGES_REQUESTED** [review]
Love it, very nice PR 🎉 

What was your motivation to create a dedicated `make` target for this? Why don't we just always register the seed when running `make operator-up`? Is "speed" our only concern?

**Comment by rfranzke** [comment]
Can you augment https://github.com/gardener/gardener/blob/master/docs/concepts/operator.md#local-development and adapt the release note with a bit more information (/pointer to the docs) for devs?

File: pkg/controller/service/reconciler.go [hunk 12]
```diff
@@ -66,29 +68,57 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
-		case key == keyIstioIngressGatewayZone0 && servicePort.Name == "tcp":
-			service.Spec.Ports[i].NodePort = nodePortIstioIngressGatewayZone0
-			ip = r.Zone0IP
-		case key == keyIstioIngressGatewayZone1 && servicePort.Name == "tcp":
-			service.Spec.Ports[i].NodePort = nodePortIstioIngressGatewayZone1
-			ip = r.Zone1IP
-		case key == keyIstioIngressGatewayZone2 && servicePort.Name == "tcp":
-			service.Spec.Ports[i].NodePort = nodePortIstioIngressGatewayZone2
-			ip = r.Zone2IP
+		}
```
Comment (ScheererJ) [inline pkg/controller/service/reconciler.go]: Can you please share the reasoning and maybe even put it into a comment why this is done? It looks fairly strange to check in the istio ingress gateway case for the virtual garden istio ingress gateway. Therefore, I would expect at least a comment explaining why something needs to be done.

File: pkg/controller/service/reconciler.go [hunk 13]
```diff
@@ -66,29 +68,57 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+			// Setting external traffic policy to local mitigates the issue for multi-node setups, e.g. for gardener-operator.
+			service.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyLocal
+		}
+		if err := r.Client.Get(ctx, keyVirtualGardenIstioIngressGateway, ptr.To(corev1.Service{})); err != nil {
+			if apierrors.IsNotFound(err) {
+				if nodes, err := r.getNodesInternalIPs(ctx); err != nil {
+					return reconcile.Result{}, err
+				} else {
+					ips = append(ips, nodes...)
+				}
```
Comment (ScheererJ) [inline pkg/controller/service/reconciler.go]: Drop the `else` after an `if` block which always returns.

```suggestion
				}
				ips = append(ips, nodes...)
```

File: example/provider-local/managedseeds/shoot-managedseed.yaml [hunk 14]
```diff
@@ -28,3 +28,5 @@ spec:
       serializeImagePulls: false
       registryPullQPS: 10
       registryBurst: 20
+    verticalPodAutoscaler:
+      enabled: true
```
Comment (ScheererJ) [inline example/provider-local/managedseeds/shoot-managedseed.yaml:33]: Why was this added? Is this required?

File: Makefile [hunk 15]
```diff
@@ -302,13 +302,11 @@ kind-operator-up: $(KIND) $(KUBECTL) $(HELM) $(YQ) $(KUSTOMIZE)
...
 		--path-cluster-values $(REPO_ROOT)/example/gardener-local/kind/operator/values.yaml \
 		--with-lpp-resize-support $(DEV_SETUP_WITH_LPP_RESIZE_SUPPORT)
-	mkdir -p $(REPO_ROOT)/dev/local-backupbuckets/gardener-operator
+
 kind-operator-down: $(KIND)
 	./hack/kind-down.sh \
 		--cluster-name $(CLUSTER_NAME) \
 		--path-kubeconfig $(KIND_KUBECONFIG)
-	# We need root privileges to clean the backup bucket directory, see https://github.com/gardener/gardener/issues/6752
-	docker run --user root:root -v $(REPO_ROOT)/dev/local-backupbuckets:/dev/local-backupbuckets alpine rm -rf /dev/local-backupbuckets/gardener-operator
```
Comment (ScheererJ) [inline Makefile]: What is the impact of moving this into the `kind-down.sh` script behind an environment variable? Will it no longer be executed per default?
I guess it solves the linux users' problem with backup bucket cleanup, but what is the impact for everybody else?

Please share your thoughts.

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for making it easily possible to work with seeds/shoots in the local operator setup.

What do you think about making it easier to start, e.g. by either joining the `make` targets or adding make dependencies? Then, it would be more similar to `make kind-up gardener-up`.

File: hack/operator-garden-seed-down.sh (hunk 7 above)
Comment (oliver-goetz) [reply hack/operator-garden-seed-down.sh]: Nope, I deleted the statement 😄

File: pkg/provider-local/webhook/dnsconfig/mutator.go (hunk 9 above)
Comment (oliver-goetz) [reply pkg/provider-local/webhook/dnsconfig/mutator.go]: It is added for machine pods only now. Blackbox exporter already has such a label.

File: example/provider-local/managedseeds/shoot-managedseed.yaml (hunk 14 above)
Comment (oliver-goetz) [reply example/provider-local/managedseeds/shoot-managedseed.yaml:33]: Yes, VPA has to be enabled on a shoot to create a managed seed ([ref](https://github.com/gardener/gardener/blob/b38df0e70183a04a2810f0d71271ca94c2b7e65f/plugin/pkg/managedseed/validator/admission.go#L203-L205)).

File: Makefile (hunk 15 above)
Comment (oliver-goetz) [reply Makefile]: Yes, it does not look like a real fix, but only leaves orphaned backup bucket behind.
I dropped the commit (@robertvolkmann).

File: Makefile (hunk 1 above)
Comment (oliver-goetz) [reply Makefile:303]: Please see [here](https://github.com/gardener/gardener/pull/9763#discussion_r1610414345).

File: charts/gardener/operator/templates/deployment.yaml (hunk 4 above)
Comment (oliver-goetz) [reply charts/gardener/operator/templates/deployment.yaml]: This is the only Helm chart where we added `hostAliases` and we could not find any other use case for it.
Is there a concrete purpose you have in mind?

**Comment by oliver-goetz** [comment]
> Love it, very nice PR 🎉

Thanks 😄 

> 
> What was your motivation to create a dedicated `make` target for this? Why don't we just always register the seed when running `make operator-up`? Is "speed" our only concern?

e2e tests of gardener-operator already use `make operator-up` ([ref](https://github.com/gardener/gardener/blob/fa76e5c4cb4f7c300e2879654fd682e791d447d3/hack/ci-e2e-kind-operator.sh#L24)). Those tests create and delete their own gardens. We have to create a new make target (or a parameter for `operator-up`)  in order to keep it that way.

File: example/provider-local/managedseeds/shoot-managedseed.yaml (hunk 14 above)
Comment (rfranzke) [reply example/provider-local/managedseeds/shoot-managedseed.yaml:33]: Isn't this the default anyways for `Shoot`s? 👀

File: charts/gardener/operator/templates/deployment.yaml (hunk 4 above)
Comment (rfranzke) [reply charts/gardener/operator/templates/deployment.yaml]: No, but that doesn't mean that nobody else has one :) I was just wondering why this was removed since it doesn't "hurt" to have this configurable.

**Comment by ScheererJ** [comment]
> > What was your motivation to create a dedicated `make` target for this? Why don't we just always register the seed when running `make operator-up`? Is "speed" our only concern?
> 
> e2e tests of gardener-operator already use `make operator-up` ([ref](https://github.com/gardener/gardener/blob/fa76e5c4cb4f7c300e2879654fd682e791d447d3/hack/ci-e2e-kind-operator.sh#L24)). Those tests create and delete their own gardens. We have to create a new make target (or a parameter for `operator-up`) in order to keep it that way.

Would it be an option to make the new `make` target dependent on `operator-up` so that you can still just use two targets to get everything up?

**Review by ScheererJ — COMMENTED** [review]
/lgtm

File: charts/gardener/operator/templates/deployment.yaml (hunk 4 above)
Comment (oliver-goetz) [reply charts/gardener/operator/templates/deployment.yaml]: I added it again

**Comment by jguipi** [comment]
Might be important to update this documentation as well when the migration to the `gardener-operator` is complete. 

https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md

**Comment by oliver-goetz** [comment]
> Might be important to update this documentation as well when the migration to the `gardener-operator` is complete.
> 
> https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md

There is already a new section in [./docs/deployment/getting_started_locally.md](https://github.com/gardener/gardener/blob/3ef6f39655666d83021491565a44e5324fb15eaf/docs/deployment/getting_started_locally.md#alternative-way-to-set-up-garden-and-seed-leveraging-gardener-operator) which is linked in the file you mentioned.

File: docs/deployment/getting_started_locally.md [hunk 16]
```diff
@@ -331,6 +331,42 @@ make kind2-down
 make kind-down
 ```
 
+## Alternative way to set up Garden and Seed leveraging `gardener-operator`
+
+Instead of starting Garden and Seed via `make kind-up gardener-up` you can also use `gardener-operator` to create your local dev landscape.
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
Instead of starting Garden and Seed via `make kind-up gardener-up`, you can also use `gardener-operator` to create your local dev landscape.  
```

File: docs/deployment/getting_started_locally.md [hunk 17]
```diff
@@ -331,6 +331,42 @@ make kind2-down
 make kind-down
 ```
 
+## Alternative way to set up Garden and Seed leveraging `gardener-operator`
+
+Instead of starting Garden and Seed via `make kind-up gardener-up` you can also use `gardener-operator` to create your local dev landscape.  
+In this setup virtual garden cluster ash its own loadbalancer, so you have to create an own DNS entry in your `/etc/hosts`.
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
In this setup, virtual garden cluster has its own load balancer, so you have to create an own DNS entry in your `/etc/hosts`.
```

File: docs/deployment/getting_started_locally.md [hunk 18]
```diff
@@ -331,6 +331,42 @@ make kind2-down
...
+```shell
+cat <<EOF | sudo tee -a /etc/hosts
+
+# Manually created to access local Gardener virtual garden cluster.
+# TODO: Remove this again when the virtual garden cluster access is no longer required.
+127.0.0.3 api.virtual-garden.local.gardener.cloud
+EOF
+```
+
+You can bring up `gardener-operator` with this command.
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
You can bring up `gardener-operator` with this command:
```

File: docs/deployment/getting_started_locally.md [hunk 19]
```diff
@@ -331,6 +331,42 @@ make kind2-down
...
+
+```shell
+make kind-operator-up operator-up
+```
+
+Then you can create your local `Garden` and install `gardenlet` into the KinD cluster with:
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
Afterwards, you can create your local `Garden` and install `gardenlet` into the KinD cluster with this command:
```

File: docs/deployment/getting_started_locally.md [hunk 20]
```diff
@@ -331,6 +331,42 @@ make kind2-down
...
+
+```shell
+make operator-seed-up
+```
+
+You find the kubeconfig for the KinD cluster at `./example/gardener-local/kind/operator/kubeconfig`. The one for the virtual garden is accessible at `./example/operator/virtual-garden/kubeconfig`.
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
You find the kubeconfig for the KinD cluster at `./example/gardener-local/kind/operator/kubeconfig`.
The one for the virtual garden is accessible at `./example/operator/virtual-garden/kubeconfig`.  
```

File: docs/deployment/getting_started_locally.md [hunk 21]
```diff
@@ -331,6 +331,42 @@ make kind2-down
...
+```
+
+Then you can create your local `Garden` and install `gardenlet` into the KinD cluster with:
+
+```shell
+make operator-seed-up
+```
+
+You find the kubeconfig for the KinD cluster at `./example/gardener-local/kind/operator/kubeconfig`. The one for the virtual garden is accessible at `./example/operator/virtual-garden/kubeconfig`.  
+Please use this command to tear down your environment.
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
Please use this command to tear down your environment:
```

File: docs/deployment/getting_started_locally.md [hunk 22]
```diff
@@ -331,6 +331,42 @@ make kind2-down
...
+
+```shell
+make kind-operator-down
+```
+
+This setup supports creating shoots and managed seeds the same way as explained in the previous chapters. However, the development and debugging setups and the e2e tests are not working yet.
+
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]:
```suggestion
```

File: skaffold.yaml [hunk 23]
```diff
@@ -848,12 +848,7 @@ deploy:
               # TODO(timebertt): drop this after v1.96 has been released
               - bash
               - -ec
-              - |
-                until kubectl get controllerdeployments.v1.core.gardener.cloud >/dev/null; do 
-                  echo "Waiting for controllerdeployments.v1.core.gardener.cloud to be available..."
-                  sleep 5
-                done
-                echo "controllerdeployments.v1.core.gardener.cloud available"
+              - "until kubectl get controllerdeployments.v1.core.gardener.cloud >/dev/null; do \n  echo \"Waiting for controllerdeployments.v1.core.gardener.cloud to be available...\"\n  sleep 5\ndone\necho \"controllerdeployments.v1.core.gardener.cloud available\"\n"
```
Comment (rfranzke) [inline skaffold.yaml]: Why did you change this? 👀

**Review by rfranzke — COMMENTED** [review]
Does it make sense to add a simple e2e test for this scenario? We could just set it up, maybe create a basic shoot and make it a `ManagedSeed`, and then tear it down again. WDYT?

File: skaffold.yaml (hunk 23 above)
Comment (oliver-goetz) [reply skaffold.yaml]: :ghost:

**Comment by oliver-goetz** [comment]
I added the e2e tests `pull-gardener-e2e-kind-operator-seed` and `ci-gardener-e2e-kind-operator-seed` in https://github.com/gardener/ci-infra/pull/1814 and started the first one manually for this PR.

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
